packages/expo-crypto/android/src/main/java/expo/modules/crypto/aes/AesCryptoModule.kt

```diff
@@ -0,0 +1,167 @@
+import expo.modules.kotlin.apifeatures.EitherType
+import expo.modules.kotlin.exception.Exceptions
+import expo.modules.kotlin.modules.Module
+import expo.modules.kotlin.modules.ModuleDefinition
+import expo.modules.kotlin.types.Either
+import java.nio.ByteBuffer
+import java.security.SecureRandom
+import javax.crypto.Cipher
+import javax.crypto.spec.GCMParameterSpec
+
+@OptIn(EitherType::class)
+typealias BinaryInput = Either<ByteArray, String>
+
+@OptIn(EitherType::class)
+class AesCryptoModule : Module() {
+  private val rng: SecureRandom by lazy { SecureRandom() }
+
+  override fun definition() = ModuleDefinition {
+    Name("ExpoCryptoAES")
+
+    AsyncFunction("encryptAsync", this@AesCryptoModule::encrypt)
+    AsyncFunction("decryptAsync", this@AesCryptoModule::decrypt)
+
+    Class("EncryptionKey", EncryptionKey::class) {
+      Constructor {
+        throw Exceptions.IllegalArgument("EncryptionKey constructor cannot be used directly")
+      }
+      StaticAsyncFunction("generate", this@AesCryptoModule::generateKey)
+      StaticAsyncFunction("import", this@AesCryptoModule::importKey)
+
+      AsyncFunction("bytes") { key: EncryptionKey -> key.bytes }
+      AsyncFunction("encoded") { key: EncryptionKey, encoding: KeyEncoding ->
+        key.bytes.encoded(encoding)
+      }
+
+      Property("size") { key: EncryptionKey -> key.keySize }
+    }
+
+    Class("SealedData", SealedData::class) {
+      Constructor {
+        throw Exceptions.IllegalArgument("SealedData constructor cannot be used directly")
+      }
+      StaticFunction("fromParts", this@AesCryptoModule::sealedDataFromParts)
+      StaticFunction("fromCombined") { combined: ByteArray, config: SealedDataConfig? ->
+        val config = config ?: SealedDataConfig()
+        SealedData(config, content = combined)
+      }
+
+      AsyncFunction("iv") { sealedData: SealedData, format: DataFormat? ->
+        sealedData.ivBytes.formatted(format)
+      }
+      AsyncFunction("tag") { sealedData: SealedData, format: DataFormat? ->
+        sealedData.tagBytes.formatted(format)
+      }
+      AsyncFunction("combined") { sealedData: SealedData, format: DataFormat? ->
+        sealedData.combinedArray.formatted(format)
+      }
+      AsyncFunction("ciphertext") { sealedData: SealedData, options: CiphertextOptions? ->
+        val (includeTag, outputFormat) = options ?: CiphertextOptions()
+        sealedData.ciphertextBytes(withTag = includeTag).formatted(outputFormat)
+      }
+
+      Property("combinedSize") { sealedData -> sealedData.combinedSize }
+      Property("ivSize") { sealedData -> sealedData.ivSize }
+      Property("tagSize") { sealedData -> sealedData.tagSize }
+    }
+  }
+
+  private fun generateKey(size: KeySize?): EncryptionKey {
+    return EncryptionKey(size ?: KeySize.AES256)
+  }
+
+  @OptIn(ExperimentalStdlibApi::class)
+  private fun importKey(input: Either<ByteArray, String>, encoding: KeyEncoding?): EncryptionKey {
```

Comment:
I would move those functions into the DSL lambda rather than using their references.